lib/posix-netlink/netlink.c

```diff
@@ -0,0 +1,339 @@
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+
+	if (*addr_len < sizeof(*nl_addr)) {
+		/* addr_len is too small, only inform about needed size */
+		goto out;
+	}
+
+	nl_addr = (struct sockaddr_nl *)addr;
+	nl_addr->nl_family = AF_NETLINK;
+	nl_addr->nl_pid = nl_ctx_pid(nl_ctx);
+	nl_addr->nl_pad = 0;
+	nl_addr->nl_groups = nl_ctx->nl_groups;
+
+out:
+	*addr_len = sizeof(*nl_addr);
+	return 0;
+}
+
+static inline int nl_handle(struct nl_ctx *nl_ctx, const void *buf, size_t len)
+{
+	const struct nlmsghdr *nlh;
+	size_t tmp_len;
+	int ret = 0;
+
+	UK_ASSERT(nl_ctx);
+	UK_ASSERT(nl_ctx->drv);
+	UK_ASSERT(nl_ctx->drv->ops);
+	UK_ASSERT(nl_ctx->drv->ops->handle);
+
+	uk_pr_debug("Handle incoming netlink msg\n");
+	if (!buf || len < sizeof(struct nlmsghdr))
+		return -EINVAL;
+
+	/* sanity check multi-messages */
+	tmp_len = len;
+	for (nlh = (const struct nlmsghdr *)buf;
+	     NLMSG_OK(nlh, tmp_len);
+	     nlh = NLMSG_NEXT(nlh, tmp_len)) {
+		if (tmp_len < nlh->nlmsg_len) {
+			uk_pr_debug("Sanity check failed\n");
+			return -EINVAL;
+		}
+	}
+
+	/* forward message by message */
+	for (nlh = (const struct nlmsghdr *)buf;
+	     NLMSG_OK(nlh, len);
+	     nlh = NLMSG_NEXT(nlh, len)) {
+		uk_pr_debug("Call handler for msg %p (driver %s)\n",
+			    nlh, nl_ctx->drv->libname);
+		ret = nl_ctx->drv->ops->handle(nl_ctx, nlh);
+	}
+
+	return ret;
+}
+
+static ssize_t nl_sendto(posix_sock *s,
+			 const void *buf, size_t len, int flags __unused,
+			 const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+	struct sockaddr_nl *nl_addr;
+	int ret;
+
+	if (addrlen < sizeof(*nl_addr))
+		return -EINVAL;
+
+	nl_addr = (struct sockaddr_nl *)dest_addr;
+	if (unlikely(nl_addr->nl_family != AF_NETLINK || nl_addr->nl_pid != 0))
+		return -EHOSTUNREACH;
+
+	ret = nl_handle(nl_ctx, buf, len);
+	if (unlikely(ret < 0))
+		return -EINVAL;
+	return len;
+}
+
+static size_t copy2iovec(struct iovec iovec[], size_t iovec_len,
+			 const void *src, size_t len)
+{
+	size_t left = len;
+	size_t cpylen;
+	size_t i = 0;
+
+	for (i = 0; i < iovec_len && left; ++i) {
+		cpylen = MIN(iovec[i].iov_len, left);
+		memcpy(iovec[i].iov_base, src, cpylen);
+
+		left -= cpylen;
+		src  += cpylen;
+	}
+
+	return (len - left);
+}
+
+static ssize_t nl_recvmsg(posix_sock *s, struct msghdr *msg,
+			  int flags __unused)
+{
+	struct uk_streambuf *nlbuf;
+	struct sockaddr_nl *nl_addr;
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+	size_t cpylen;
+
+	uk_pr_debug("Picking up packet from netlink mbox\n");
+	uk_mbox_recv(nl_ctx->nl_recvqueue, (void **)&nlbuf);
+	if (unlikely(!nlbuf))
+		return -EIO;
+
+	cpylen = copy2iovec(msg->msg_iov, msg->msg_iovlen,
+			    nlbuf_data(nlbuf), nlbuf_len(nlbuf));
+	nlbuf_free(nlbuf);
+	uk_pr_debug("Message received\n");
+
+	/* Fill name structure */
+	if (msg->msg_name && msg->msg_namelen == sizeof(struct sockaddr_nl)) {
+		nl_addr = msg->msg_name;
+		nl_addr->nl_family = AF_NETLINK;
+		nl_addr->nl_pid = 0;
```

Comment:
I take it that for now we only support process to kernel communication hence pid is always set to 0?